client/app/components/queries/editor-components/editorComponents.js

```diff
@@ -0,0 +1,24 @@
+import { isArray, isNil, each } from "lodash";
+import { useMemo } from "react";
+
+const componentsRegistry = new Map();
+
+export function registerEditorComponent(componentName, component, dataSourceTypes) {
+  if (isNil(dataSourceTypes)) {
+    dataSourceTypes = [null]; // use `null` entry for the default set of components
+  }
+
+  if (!isArray(dataSourceTypes)) {
+    dataSourceTypes = [dataSourceTypes];
+  }
+
+  each(dataSourceTypes, dataSourceType => {
+    componentsRegistry.set(dataSourceType, { ...componentsRegistry.get(dataSourceType), [componentName]: component });
+  });
+}
+
+export function useEditorComponents(dataSourceType) {
```

Comment:
Is this a hook so we can use `useMemo`? If it is, why do we need `useMemo` won't it always return the same object (the component class) for the same input?

Reply:
Mostly to use `useMemo` indeed, the plus is that it seems the "React way" nowadays.

Well, while it should return the same input, objects and arrays always hold different comparison values when you instantiate them, which can trigger updates unnecessarily elsewhere. To make things "safer" I usually define a `useMemo` for non-basic variable types.

Reply:
> the plus is that it seems the "React way" nowadays.

Our code is Javascript first, React second. This to mean that we should use Javascript features first (simple functions) when possible.

> while it should return the same input, objects and arrays always hold different comparison values when you instantiate them, which can trigger updates unnecessarily elsewhere.

But the way these values are used won't make React re-render when they change, so it doesn't matter that we create a new object every time this function is called.

It might seem small, but needlessly adding `useMemo` (and other optimizations) in the code makes the code harder to follow and less performant.

Reply:
> But the way these values are used won't make React re-render when they change, so it doesn't matter that we create a new object every time this function is called.

I thought about it for a while, and you are right, for the way we will use it (destructuring) it shouldn't affect. I was too focused looking only at the function returning an object 😅.

> It might seem small, but needlessly adding useMemo (and other optimizations) in the code makes the code harder to follow and less performant.

I agree and you can be sure that I try not to use it needlessly, the thing that usually triggers me to use a hook instead of a simple JS function is when something can be potentially forgotten to receive the optimization, thus maybe invalidating other existing optimizations in the component. E.g.: a transform function for an user object (`userObject => anotherObject`), it's simple, but the different address can trigger unnecessary updates.